chainer/functions/normalization/group_normalization.py

```diff
@@ -1,10 +1,348 @@
+        batch_size, channels = orig_shape[:2]
+        groups = self.groups
+        reduced_shape = (batch_size * groups, -1)
+        x = x.reshape(reduced_shape)
+
+        self.mean = x.mean(axis=1)
+        x_hat = x - self.mean[:, None]
+        var = (x_hat * x_hat).mean(axis=1)
+        var += self.eps
+        self.inv_std = var
+        del var
+        xp.sqrt(self.inv_std, out=self.inv_std)
+        xp.reciprocal(self.inv_std, out=self.inv_std)
+        x_hat *= self.inv_std[:, None]
+
+        y = x_hat.reshape((batch_size, channels, -1))
+        y *= gamma[:, None]
+        y += beta[:, None]
+
+        y = y.reshape(orig_shape)
+        return y,
+
+    def forward_cudnn(self, inputs):
+        if self.eps < libcudnn.CUDNN_BN_MIN_EPSILON:
+            raise RuntimeError(
+                'cuDNN does not allow an eps value '
+                'less than {}.'.format(libcudnn.CUDNN_BN_MIN_EPSILON))
+
+        self.retain_inputs((0, 1))
+        x, gamma, beta = inputs
+        xp = cuda.cupy
+
+        orig_shape = x.shape
+        batch_size, channels = orig_shape[:2]
+        groups = self.groups
+        cudnn_shape = (1, batch_size * groups, -1, 1)
+        x = x.reshape(cudnn_shape)
+
+        with cuda.get_device_from_array(x):
+            dummy_beta = xp.zeros(batch_size * groups, dtype=x.dtype)
+            self.dummy_gamma = xp.ones_like(dummy_beta)
+            self.mean = xp.empty_like(dummy_beta)
+            self.inv_std = xp.empty_like(dummy_beta)
+        x_hat = cudnn.batch_normalization_forward_training(
+            x, self.dummy_gamma, dummy_beta, dummy_beta, dummy_beta,
+            self.mean, self.inv_std, self.eps, 1.0,
+            True, libcudnn.CUDNN_BATCHNORM_SPATIAL,
+            configuration.config.debug)
+
+        y = x_hat.reshape((batch_size, channels, -1))
+        cuda.elementwise(
+            'T gamma, T beta', 'T y',
+            'y = y * gamma + beta',
+            'groupnorm_y')(gamma[:, None], beta[:, None], y)
+
+        y = y.reshape(orig_shape)
+        return y,
+
+    def backward(self, indexes, grad_outputs):
+        x, gamma = self.get_retained_inputs()
+        gy, = grad_outputs
+
+        orig_shape = x.shape
+        batch_size = orig_shape[0]
+        groups = self.groups
+        reduced_shape = (batch_size * groups, -1)
+        x = chainer.functions.reshape(x, reduced_shape)
+
+        x_hat, = _XHat(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma).apply((x,))
+        gx_hat, ggamma, gbeta = _GradHelper().apply((x_hat, gamma, gy))
+        gx, = _XHatGrad(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma, x_hat.array).apply((x, gx_hat))
+
+        gx = gx.reshape(orig_shape)
+        return gx, ggamma, gbeta
+
+
+class _GradHelper(function_node.FunctionNode):
+
+    def forward(self, inputs):
+        self.retain_inputs((0, 1, 2))
+        x_hat, gamma, gy = inputs
+        xp = backend.get_array_module(x_hat)
```

Comment:
A detail but let's declare `xp` closer to where it's used. Also, it could be removed since it's only used once.